pkg/sarif/sarif.go

```diff
@@ -261,6 +283,180 @@ func severityToLevel(severity v1.Severity) string {
+	var (
+		targets   = make([]string, 0)
+		parseErrs error
+	)
+
+	// We can have cases were both locations are defined but, in these cases,
+	// the physical location doesn't make much sense, so we just should leverage the logical one.
+	if isLogicalLocation(loc) {
+		tts, err := parseLogicalTargets(loc, lang...)
+		if err != nil {
+			parseErrs = errors.Join(parseErrs, err)
+		}
+		targets = append(targets, tts...)
+	} else if isPhysicalLocation(loc) {
+		tt, err := parsePhysicalTarget(loc)
+		if err != nil {
+			parseErrs = errors.Join(parseErrs, err)
+		}
+		targets = append(targets, tt)
 	}
+
+	return targets, parseErrs
+}
+
+func isPhysicalLocation(loc *sarif.Location) bool {
+	return loc.PhysicalLocation != nil
+}
+
+// parsePhysicalTarget parses a sarif physical location target.
+// We prefix the target with 'file://' so that we know that it's a physical target.
+// We add start and end lines information if provided.
+// In the case of missing start or end line, we default the undefined one to the same value of the defined one
+// because it usually means that the finding affects one line only.
+// It can be possible for a target URI to contain a pURL. In that case we simply return that.
+func parsePhysicalTarget(loc *sarif.Location) (string, error) {
+	if loc.PhysicalLocation == nil {
+		return "", nil
+	}
+
+	var (
+		target    string
+		phyLoc    = loc.PhysicalLocation
+		targetURI = phyLoc.ArtifactLocation.URI
+	)
+
+	if targetURI == nil || *targetURI == "" {
+		return "", errors.New("target URI is empty")
+	}
+
+	// This means that it's a purl in a physical path.
+	if _, err := packageurl.FromString(*targetURI); err == nil {
+		return *targetURI, nil
+	}
+
+	// Safety check to make sure that we don't end up with malformed targets if
+	// the upstream tool is already doing this.
+	if !strings.Contains(*targetURI, "file://") {
+		target = fmt.Sprintf("file://%s", *targetURI)
+	}
+
+	if phyLoc.Region != nil {
+		var (
+			isStartLineDefined = phyLoc.Region.StartLine != nil
+			isEndLineDefined   = phyLoc.Region.EndLine != nil
+		)
+
+		switch {
+		case isStartLineDefined && isEndLineDefined:
+			target = fmt.Sprintf(
+				"%s:%d-%d",
+				target,
+				*phyLoc.Region.StartLine,
+				*phyLoc.Region.EndLine,
+			)
+		case isStartLineDefined && !isEndLineDefined:
+			target = fmt.Sprintf(
+				"%s:%d-%d",
+				target,
+				*phyLoc.Region.StartLine,
+				*phyLoc.Region.StartLine,
+			)
+		case !isStartLineDefined && isEndLineDefined:
+			target = fmt.Sprintf(
+				"%s:%d-%d",
+				target,
+				*phyLoc.Region.EndLine,
+				*phyLoc.Region.EndLine,
+			)
+		}
+	}
+
+	return target, nil
+}
+
+func isLogicalLocation(loc *sarif.Location) bool {
+	return len(loc.LogicalLocations) > 0
+}
+
+// parseLogicalTargets parses all the targets found in the logical locations.
+// A logical target must be a valid pURL.
+func parseLogicalTargets(loc *sarif.Location, lang ...ExtraContextLanguage) ([]string, error) {
```

Comment:
I don't understand this choice, why have the language be a slice of arbitrary length? this could easily be invoked with `parseLogicalTargets(someLocation, "javascript", "python")` and it would not throw an error. since the code can only handle one language every time we should allow only one language to be passed.

this piece of code would be simplified if it was a method of a struct that could hold the context of the code.
for example we could have `sarif.Parser` and define a method called `WithContext(someLanguage string)` to add context when we have it so that we can reduce the number of parameters needed when the publicly available methods are invoked instead of having one parameter percolate through three functions (`ToSmithy` -> `parseOut` -> `parseTargets`)

Reply:
Right now it uses the first item in the slice, if defined

I did it to not break the existing contract for the `ToSmithy` parser

I thought of refactoring the code more heavily to do what you suggested but I'm not sure if it's worth the investment since we'll soon rework on this in the OSS

If you feel strongly about it I can make more changes

Reply:
I looked really quick into how many components are using the sarif parser and it (https://github.com/search?q=repo%3Asmithy-security%2Fsmithy%20sarif.ToSmithy&type=code) looks there is 5-6 out there which makes me question how accurate they are at parsing results, but that's a discussion for another day. for now I don't think it will be a huge amount of changes if we add a language parameter to the `ToSmithy` function and just pass an empty string to the components that we are not interested atm. this will also not add too many changes to the current PR
I guess this piece of code will be completely refactored in order to integrate with the new SKD anyway

Reply:
Sounds good - I passed the language and used `Unspecified` when not needed

The code handles both empty val and unspecified to understand that it should not be used.